flow/display_list.h

```diff
@@ -0,0 +1,427 @@
+  V(SetBlendMode)                   \
+                                    \
+  V(SetFilterQuality)               \
+                                    \
+  V(SetShader)                      \
+  V(ClearShader)                    \
+  V(SetColorFilter)                 \
+  V(ClearColorFilter)               \
+  V(SetImageFilter)                 \
+  V(ClearImageFilter)               \
+                                    \
+  V(ClearMaskFilter)                \
+  V(SetMaskFilter)                  \
+  V(SetMaskBlurFilterNormal)        \
+  V(SetMaskBlurFilterSolid)         \
+  V(SetMaskBlurFilterOuter)         \
+  V(SetMaskBlurFilterInner)         \
+                                    \
+  V(Save)                           \
+  V(SaveLayer)                      \
+  V(SaveLayerBounds)                \
+  V(Restore)                        \
+                                    \
+  V(Translate)                      \
+  V(Scale)                          \
+  V(Rotate)                         \
+  V(Skew)                           \
+  V(Transform2x3)                   \
+  V(Transform3x3)                   \
+                                    \
+  V(ClipRect)                       \
+  V(ClipRRect)                      \
+  V(ClipPath)                       \
+                                    \
+  V(DrawPaint)                      \
+  V(DrawColor)                      \
+                                    \
+  V(DrawLine)                       \
+  V(DrawRect)                       \
+  V(DrawOval)                       \
+  V(DrawCircle)                     \
+  V(DrawRRect)                      \
+  V(DrawDRRect)                     \
+  V(DrawArc)                        \
+  V(DrawPath)                       \
+                                    \
+  V(DrawPoints)                     \
+  V(DrawLines)                      \
+  V(DrawPolygon)                    \
+  V(DrawVertices)                   \
+                                    \
+  V(DrawImage)                      \
+  V(DrawImageRect)                  \
+  V(DrawImageNine)                  \
+  V(DrawImageLattice)               \
+  V(DrawAtlas)                      \
+  V(DrawAtlasColored)               \
+  V(DrawAtlasCulled)                \
+  V(DrawAtlasColoredCulled)         \
+                                    \
+  V(DrawSkPicture)                  \
+  V(DrawDisplayList)                \
+  V(DrawTextBlob)                   \
+  /* V(DrawShadowRec) */            \
+                                    \
+  V(DrawShadow)
+
+#define DL_OP_TO_ENUM_VALUE(name) name,
+enum DisplayListOpType { FOR_EACH_DISPLAY_LIST_OP(DL_OP_TO_ENUM_VALUE) };
+#undef DL_OP_TO_ENUM_VALUE
+
+class Dispatcher;
+class DisplayListBuilder;
+
+// The base class that contains a sequence of rendering operations
+// for dispatch to a Dispatcher. These objects must be instantiated
+// through an instance of DisplayListBuilder::build().
+class DisplayList : public SkRefCnt {
```

Comment:
Why are we extending `SkRefCnt`, as opposed to using `fml::RefPtr` or `std::shared_ptr`?

Reply:
For consistency with SkPicture that it is replacing? That was probably my default reasoning. Also, every other complex structure included in the DL by reference supports SkRefCnt which might be helpful if we ever want to do some other bulk operations on the DL contents.

I found in testing that sk_sp<Foo> can be compared using a bulk compare since it doesn't actually wrap the object. Basically: `sk_sp<Foo> == &Foo` so bulk comparing across a bunch of sk_sp<> pointers is equivalent to bulk comparing across a bunch of raw pointers to the objects themselves. I'm not sure about the others, but since std::shared_ptr can work on any type then I'd assume that it is a wrapper which would require a lot more care in equals comparisons.

Reply:
SGTM. As a nitpick, might worth documenting that reasoning here, since this will be the only SkRefCnt we define.